packages/unmock-node/src/backend/index.ts

```diff
@@ -24,58 +34,148 @@ async function handleRequestAndResponse(
     const serializedResponse: ISerializedResponse | undefined = createResponse(
       serializedRequest,
     );
+
     if (serializedResponse === undefined) {
-      throw Error(constants.MESSAGE_FOR_MISSING_MOCK);
+      // TODO Handle this properly
+      debugLog("No match found, emitting error");
+      clientRequest.emit("error", Error("No matching template found"));
+      return;
     }
     respondFromSerializedResponse(serializedResponse, res);
   } catch (err) {
-    // TODO Emit an error in the corresponding client request instead?
-    const errorResponse: ISerializedResponse = {
-      body: err.message,
-      statusCode: constants.STATUS_CODE_FOR_ERROR,
-    };
-    respondFromSerializedResponse(errorResponse, res);
+    clientRequest.emit("error", Error(`unmock error: ${err.message}`));
   }
 }
 
 export interface INodeBackendOptions {
   servicesDirectory?: string;
 }
 
+interface IBypassableSocket extends net.Socket {
+  bypass: () => void;
+}
+
 const nodeBackendDefaultOptions: INodeBackendOptions = {};
+const UNMOCK_INTERNAL_HTTP_HEADER = "x-unmock-req-id";
 
-let mitm: any;
 export default class NodeBackend implements IBackend {
   private readonly config: INodeBackendOptions;
+  private clientRequests: {
+    [requestId: string]: {
+      clientRequest: ClientRequest;
+    };
+  } = {};
+  private origOnSocket?: (socket: net.Socket) => void;
+  private mitm: any;
+
   public constructor(config?: INodeBackendOptions) {
     this.config = { ...nodeBackendDefaultOptions, ...config };
   }
 
   public initialize(options: UnmockOptions) {
-    mitm = Mitm();
-    mitm.on("connect", (socket: any, opts: any) => {
-      if (options.isWhitelisted(opts.host)) {
-        socket.bypass();
-      }
-    });
+    if (this.mitm !== undefined) {
+      this.reset();
+    }
+    this.mitm = Mitm();
+
+    const self = this;
+    this.origOnSocket = ClientRequest.prototype.onSocket;
+
+    /**
+     * When a socket is assigned to a client request, create an internal ID
+     * and add the ID in the request header. Thereby we can map
+     * the server-side incoming request to the corresponding
+     * client request and emit errors properly.
+     * Borrowed from https://github.com/FormidableLabs/yesno.
+     */
+    ClientRequest.prototype.onSocket = _.flowRight(
+      this.origOnSocket,
+      function(
+        this: ClientRequest,
+        socket: IBypassableSocket,
+      ): IBypassableSocket {
+        const requestId = uuidv4();
+        debugLog(
+          `New socket assigned to client request, assigned ID: ${requestId}`,
+        );
+        self.clientRequests[requestId] = { clientRequest: this };
+        this.setHeader(UNMOCK_INTERNAL_HTTP_HEADER, requestId);
+        return socket;
+      },
+    );
+
+    // Client-side socket connect, use to bypass connections
+    this.mitm.on("connect", (socket: IBypassableSocket, opts: RequestOptions) =>
+      this.mitmOnConnect.call(this, options, socket, opts),
```

Comment:
Why the need to use `call` with `this` instead of `this.mitmOnConnect(options, socket, opts)`?

Reply:
Because in the latter case the function `this.mitmOnConnect(options, socket, opts)` would have `this` bound by the context, not necessarily being the class instance.

Reply:
Actually it's true this does not need binding the context as it uses an arrow function and does not set the instance method as callback directly, I'll simplify this sometime.